src/transformers/models/mixtral/modeling_mixtral.py

```diff
@@ -736,28 +736,34 @@ def forward(
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
         kv_seq_len = key_states.shape[-2]
+        past_key_value = getattr(self, "past_key_value", past_key_value)
         if past_key_value is not None:
-            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
+            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)  # add what was seen
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
 
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
 
+        past_seen_tokens = kv_seq_len - key_states.shape[-2]
+        new_cache_positions = torch.arange(past_seen_tokens, past_seen_tokens + q_len, device=key_states.device)
         if past_key_value is not None:
-            cache_kwargs = {"sin": sin, "cos": cos}  # Specific to RoPE models
+            cache_kwargs = {"sin": sin, "cos": cos, "position_ids": new_cache_positions}  # Specific to RoPE models
             key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
 
         key_states = repeat_kv(key_states, self.num_key_value_groups)
         value_states = repeat_kv(value_states, self.num_key_value_groups)
 
-        if attention_mask is not None:
-            if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
-                raise ValueError(
-                    f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
-                )
+        if (
+            attention_mask is not None and not torch.all(attention_mask[..., 0] == 1) and q_len != 1
+        ):  # user defined causal mask
+            causal_mask = attention_mask[:, :, past_seen_tokens : past_seen_tokens + q_len, : key_states.shape[-2]]
```

Comment:
My benchmarking script fails here with the following traceback:

```bash
line 1040, in _update_causal_mask
    padding_mask = causal_mask[...,:mask_length].eq(0.0) * attention_mask[:, None, None, :].eq(0.0)
RuntimeError: The size of tensor a (4096) must match the size of tensor b (4097) at non-singleton dimension 3
```
This is because the casual mask is by default created with the same size as `max_positon_embeddings`. I think this assumption can be risky because many users run generation with large context (which is what I do in my benchmarking script), so the PR introduces a breaking behaviour for these users by default. IMO this should be revisited by dynamically re-allocating the `causal_mask` if we go beyond `max_positon_embeddings`. 
Currently running a simple generation with Llama2 with a context that is greater than 4096 fails with this PR, which is possible currently on main

Reply:
No again, it's not breaking if we just take care of it. 
There is a clause for this in `_update_causal_mask` which just needs to allocate a bigger mask. 
Again the overhead is None compared to main for this as we always allocate it. Quite the contrary this means that we dynamically, by stride of `max_embedding_dim` allocate the causal mask needed, instead of always re-creating it in the forward